airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/SshTunnel.java

```diff
@@ -0,0 +1,245 @@
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
+import org.bouncycastle.openssl.PEMKeyPair;
+import org.bouncycastle.openssl.PEMParser;
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+// todo (cgardens) - this needs unit tests. it is currently tested transitively via source postgres
+// integration tests.
+/**
+ * Encapsulates the connection configuration for an ssh tunnel port forward through a proxy/bastion
+ * host plus the remote host and remote port to forward to a specified local port.
+ */
+public class SshTunnel implements AutoCloseable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SshTunnel.class);
+
+  public enum Method {
+    NO_TUNNEL,
+    SSH_PASSWORD_AUTH,
+    SSH_KEY_AUTH
+  }
+
+  public static final int TIMEOUT_MILLIS = 15000; // 15 seconds
+  private final Method method;
+  private final String host;
+  private final String tunnelSshPort;
+  private final String user;
+  private final String sshkey;
+  private final String password;
+  private final String remoteDatabaseHost;
+  private final String remoteDatabasePort;
+  private final String tunnelDatabasePort;
+
+  private SshClient sshclient;
+  private ClientSession tunnelSession;
+
+  public SshTunnel(final Method method,
+                   final String host,
+                   final String tunnelSshPort,
+                   final String user,
+                   final String sshKey,
+                   final String password,
+                   final String remoteDatabaseHost,
+                   final String remoteDatabasePort,
+                   final String tunnelDatabasePort) {
+
+    Preconditions.checkNotNull(method);
+    this.method = method;
+
+    if (method.equals(Method.NO_TUNNEL)) {
+      this.host = null;
+      this.tunnelSshPort = null;
+      this.user = null;
+      this.sshkey = null;
+      this.password = null;
+      this.remoteDatabaseHost = null;
+      this.remoteDatabasePort = null;
+      this.tunnelDatabasePort = null;
+    } else {
+      Preconditions.checkNotNull(host);
+      Preconditions.checkNotNull(tunnelSshPort);
+      Preconditions.checkNotNull(user);
+      Preconditions.checkArgument(sshKey != null || password != null,
+          "SSH Tunnel was requested to be opened while it was already open.  This is a coding error.");
+      Preconditions.checkNotNull(remoteDatabaseHost);
+      Preconditions.checkNotNull(remoteDatabasePort);
+      Preconditions.checkNotNull(tunnelDatabasePort);
+
+      this.host = host;
+      this.tunnelSshPort = tunnelSshPort;
+      this.user = user;
+      this.sshkey = sshKey;
+      this.password = password;
+      this.remoteDatabaseHost = remoteDatabaseHost;
+      this.remoteDatabasePort = remoteDatabasePort;
+      this.tunnelDatabasePort = tunnelDatabasePort;
+
+      this.sshclient = createClient();
+      this.tunnelSession = openTunnel(sshclient);
```

Comment:
why are we opening a tunnel in the constructor of this object? shouldn't this either be opened explicitly or be re-entrant/a singleton?

Reply:
why? the original version was a mix between a factory and this. so i just picked one of the patterns. what does singleton or reentrant get us here?

Reply:
sorry, I mixed two concerns in my initial comment. 

1. The main unexpected behavior I'm seeing here is that we instantiate a tunnel upon creating an object. It just seemed unintuitive. I guess this is fine since it's meant to be used in a try-with-resources? 
2. I remember there was a case where we got into trouble for trying to open the tunnel twice (because read calls check or something) hence the re-entrant suggestion, so we don't create two tunnels.